enterprise/functions/src/main/java/io/crate/window/NthValueFunctions.java

```diff
@@ -0,0 +1,118 @@
+    private int seenFrameUpperBound = -1;
+    private Object resultForCurrentFrame = null;
+
+    private NthValueFunctions(FunctionInfo info, BiFunction<WindowFrameState, Input[], Integer> frameIndexSupplier) {
+        this.info = info;
+        this.frameIndexSupplier = frameIndexSupplier;
+    }
+
+    @Override
+    public FunctionInfo info() {
+        return info;
+    }
+
+    @Override
+    public Object execute(int rowIdx,
+                          WindowFrameState currentFrame,
+                          List<? extends CollectExpression<Row, ?>> expressions,
+                          Input... args) {
+        if (isNewFrame(seenFrameUpperBound, currentFrame)) {
+            seenFrameUpperBound = currentFrame.upperBoundExclusive();
+            Object[] nthRowCells = currentFrame.getRowAtIndexOrNull(frameIndexSupplier.apply(currentFrame, args));
+            if (nthRowCells == null) {
+                resultForCurrentFrame = null;
+                return null;
+            }
+
+            Row nthRowInFrame = new RowN(nthRowCells);
+            for (CollectExpression<Row, ?> expression : expressions) {
+                expression.setNextRow(nthRowInFrame);
+            }
+
+            resultForCurrentFrame = args[0].value();
+        }
+
+        return resultForCurrentFrame;
+    }
+
+    public static void register(EnterpriseFunctionsModule module) {
+        module.register(FIRST_VALUE_NAME, new BaseFunctionResolver(FuncParams.SINGLE_ANY) {
+
+            @Override
+            public FunctionImplementation getForTypes(List<DataType> dataTypes) throws IllegalArgumentException {
+                return new NthValueFunctions(
+                    new FunctionInfo(
+                        new FunctionIdent(FIRST_VALUE_NAME, dataTypes), dataTypes.get(0), FunctionInfo.Type.WINDOW),
+                    (frame, inputs) -> 0
+                );
+            }
+        });
+        module.register(LAST_VALUE_NAME, new BaseFunctionResolver(FuncParams.SINGLE_ANY) {
+
+            @Override
+            public FunctionImplementation getForTypes(List<DataType> dataTypes) throws IllegalArgumentException {
+                return new NthValueFunctions(
+                    new FunctionInfo(
+                        new FunctionIdent(LAST_VALUE_NAME, dataTypes), dataTypes.get(0), FunctionInfo.Type.WINDOW),
+                    (frame, inputs) -> frame.size() - 1
+                );
+            }
+        });
+        module.register(NTH_VALUE, new BaseFunctionResolver(FuncParams.builder(Param.ANY, Param.NUMERIC).build()) {
+
+            @Override
+            public FunctionImplementation getForTypes(List<DataType> dataTypes) throws IllegalArgumentException {
+                return new NthValueFunctions(
+                    new FunctionInfo(new FunctionIdent(NTH_VALUE, dataTypes), dataTypes.get(0), FunctionInfo.Type.WINDOW),
+                    (frame, inputs) -> ((Number) inputs[1].value()).intValue() - 1
```

Comment:
Wouldn't `nth_value(col1, null)` fail here?

Reply:
ah you're right. I thought the `Numeric` type of the second argument filters out these cases